services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java

```diff
@@ -0,0 +1,572 @@
+
+        //FIXME: Not support yet
+    }
+
+    private void rollBackAllOperations(Stack<PortStateRollback> rollbacks)
+            throws Exception {
+        while (!rollbacks.isEmpty()) {
+            rollbacks.pop().doRollback();
+        }
+    }
+
+    private HostState getHostState(String hostId) {
+        return null;
+    }
+
+    private void addPortToHost(String hostId) {
+        HostState hostState = getHostState(hostId);
+
+        //FIXME: Add port to Host
+    }
+
+    private void setDefaultValue(PortState portState) {
+        if (portState.getId() == null) {
+            portState.setId("");
+        }
+
+        if (portState.getProjectId() == null) {
+            portState.setProjectId("");
+        }
+
+        if (portState.getName() == null) {
+            portState.setName("");
+        }
+        if (portState.getDescription() == null) {
+            portState.setDescription("");
+        }
+        if (portState.getVpcId() == null) {
+            portState.setVpcId("");
+        }
+
+        if (portState.getTenantId() == null) {
+            portState.setTenantId("");
+        }
+
+        if (portState.getMacAddress() == null) {
+            portState.setMacAddress("");
+        }
+
+        if (portState.getVethName() == null) {
+            portState.setVethName("");
+        }
+
+        if (portState.getDeviceId() == null) {
+            portState.setDeviceId("");
+        }
+
+        if (portState.getDeviceOwner() == null) {
+            portState.setDeviceOwner("");
+        }
+
+        if (portState.getStatus() == null) {
+            portState.setStatus("");
+        }
+
+        if (portState.getFixedIps() == null) {
+            portState.setFixedIps(new ArrayList<>());
+        }
+
+        if (portState.getAllowedAddressPairs() == null) {
+            portState.setAllowedAddressPairs(new ArrayList<>());
+        }
+
+        if (portState.getExtraDhcpOpts() == null) {
+            portState.setExtraDhcpOpts(new ArrayList<>());
+        }
+
+        if (portState.getSecurityGroups() == null) {
+            portState.setSecurityGroups(new ArrayList<>());
+        }
+
+        if (portState.getBindingHostId() == null) {
+            portState.setBindingHostId("");
+        }
+
+        if (portState.getBindingProfile() == null) {
+            portState.setBindingProfile("");
+        }
+
+        if (portState.getBindingVnicType() == null) {
+            portState.setBindingVnicType("");
+        }
+
+        if (portState.getNetworkNamespace() == null) {
+            portState.setNetworkNamespace("");
+        }
+
+        if (portState.getDnsName() == null) {
+            portState.setDnsName("");
+        }
+
+        if (portState.getDnsAssignment() == null) {
+            portState.setDnsAssignment(new ArrayList<>());
+        }
+    }
+
+    private PortStateJson tryCreatePortState(String projectId, PortState portState, Stack<PortStateRollback> rollbacks) throws Exception {
```

Comment:
The calls into various downstream microservices remain serial and sync. 

One recommended way is to async call into multiple microservices in parallel. See the implementation of Subnet Manager for an example. 

https://github.com/futurewei-cloud/alcor/blob/a4db5c2b368e1a500f4defd62de174d375419392/services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/controller/SubnetController.java

Reply:
Okay, I will do that.

Reply:
Thanks.

Reply:
Nice work on support Async with a generic AsyncExecutor class. Like it!

Reply:
If this mechanism is approved by you, I will move it to the lib directory

Reply:
Sure. Go for it :-)